File: 백준/Silver/21318. 피아노 체조/피아노 체조.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n, q int
	fmt.Fscan(reader, &n)
	music, sum := make([]int, n), make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &music[i])
	}

	for i := 1; i < n; i++ {
		if music[i-1] > music[i] {
			sum[i] = sum[i-1] + 1
		} else {
			sum[i] = sum[i-1]
		}
	}

	fmt.Fscan(reader, &q)

	for i := 0; i < q; i++ {
		var x, y int
		fmt.Fscan(reader, &x, &y)
		if x == y {
			fmt.Fprintln(writer, 0)
		} else {
			fmt.Fprintln(writer, sum[y-1]-sum[x-1])
		}
	}

}

File: 백준/Silver/21318. 피아노 체조/피아노 체조.cc
#include <iostream>
#include <vector>

using namespace std;

int main() {
	int n, q;
	scanf("%d", &n);
	vector<int> music(n, 0), sum(n, 0);
	for (int i = 0; i < n; i++) {
		scanf("%d", &music[i]);
	}

	for (int i = 1; i < n; i++) {
		if (music[i - 1] > music[i]) {
			sum[i] = sum[i - 1] + 1;
		}
		else {
			sum[i] = sum[i - 1];
		}
	}
	scanf("%d", &q);
	for (int i = 0; i < q; i++) {
		int x, y;
		scanf("%d %d", &x, &y);
		if (x == y) {
			printf("0\n");
		}
		else {
			printf("%d\n", sum[y - 1] - sum[x - 1]);
		}
	}
	return 0;
}